### PR TITLE
Generating *.meta.js during production build

### PR DESCRIPTION
**Pull request by Giglium**
This draft PR want to add the feature of the automatic generation of the `*.meta.js files`. I'm looking for a feedback or a confirmation if this is the appropriate direction.

I'm don't really like the idea to user the `replace` function to generate the meta filenames, but is the most efficient way that I found since I prefer to avoid to generate twice the metaBlock. Any other ideas ?

In your opinion, we should introduce a flag to enable or disable the generation of these meta files?

Thank you for your feedback!

### CODE REVIEW

**Comment by Giglium** [comment]
Hi @guansss,
Do you had a chance to review this PR? 

Thanks for your time!

**Comment by guansss** [comment]
Hi, sorry for the late response! And thank you for this PR, this is a great feature to have!

I wonder why the meta file goes to `chunk.auxiliaryFiles`? I haven't used it before so I'm not sure what differences it will make.

Using `replace` to generate the meta filename is definitely okay since the js output's filename is fixed as `*.user.js`. But there is a pitfall in `.replace("user", "meta")`, where something like `"user-blocker.user.js"` will be converted to `"meta-blocker.user.js"`. It should probably be `.replace(/\.user\.js$/, ".meta.js")`.

Having a flag would be nice. I imagine something like `options.meta.file` or `options.meta.outFile`?

**Comment by Giglium** [comment]
Thanks for the review! Certainly, regex is the most robust way to perform the `replace` avoiding problems. 

For the `auxiliaryFiles` I have used this guide as a reference: https://github.com/webpack/webpack/issues/11425#issuecomment-698837871  
I understand that we should use them for any file generated during the build that is not a `.js` or `.css`.  For this reason, I used them.

For the flag, I was thinking something like `options.meta.generate`  with `true` as the default value. The `options.meta.file` or `options.meta.outFile` let me think of having the option to specify the meta filename or output file but I won't provide this possibility...or we should? 

In the next few days, I will update the PR with the modifications.

**Comment by guansss** [comment]
Thanks for the update!

In my opinion, the word "generate" fits well, but as an option, it may be a bit too broad and could represent the entire process of converting the original meta.js to the meta comment block (both in *.user.js and *.meta.js). I think at least the word "file" should be included to indicate that we are going to generate an extra file. Maybe `options.meta.generateFile`?

File: src/node/MonkeyPlugin.ts [hunk 1]
```diff
@@ -599,25 +600,29 @@ export class MonkeyPlugin {
...
+                          })),
+                        ],
+                      })
+
+                      // Generate meta file for userscript (*.meta.js)
+                      if(this.options.meta?.generate ?? true) {            
+                      const metaFile = jsFile.replace(/\.user\.js$/, ".meta.js")
+                      compilation.emitAsset(metaFile, new RawSource(metaBlock))
+                      chunk.auxiliaryFiles.add(metaFile)
+                      }
```
Comment (guansss) [inline src/node/MonkeyPlugin.ts]: Little indentation issue here.

Comment (Giglium) [reply src/node/MonkeyPlugin.ts]: My bad auto-save indenter was off

**Comment by Giglium** [comment]
Changed to `options.meta.generateFile` it's good!  I've also squashed all the commit into one!

**Review by guansss — APPROVED** [review]
Thank you!

**Pull request merged**